### ISSUE TITLE
Security issue with Highcharts 7.2.1

### ISSUE THREAD

**Issue by shakunvohra** [highcharts/highcharts#13225]
Method function at line 1302 of /js/highcharts-7.2.1/highcharts.js gets user input for
the split element. This element’s value then flows through client-side code without being properly sanitized or
validated and is eventually displayed to the user in function at line 1302 of 
js/highcharts-7.2.1/highcharts.js.This may enable a DOM XSS attack.

#### Live demo with steps to reproduce
Code Snippet
File Name /js/highcharts-7.2.1/highcharts.js
Method init: function(a, d, e, m, w, g, r) {
....
1315. this.url = (k || x) &&
p.getElementsByTagName("base").length ?
Q.location.href.split("#")[0].replace(/<[^>]*>/g,
"").replace(/([\('\)])/g, "\\$1").replace(/ /g, "%20") : "";
1316.
this.createElement("desc").add().element.appendChild(p.createTextNode("C
reated with Highcharts 7.2.1"));

#### Product version
HighCharts 7.2.1

**Comment by sebastianbochan** [comment]
Hi @shakunvohra,
Thank you for reporting about the problem.

**Comment by shakunvohra** [comment]
@sebastianbochan Do you know if HighCharts version 8 has fixed this issue?

**Comment by sebastianbochan** [comment]
Hi @shakunvohra,
As I see its still valid in the newest release: 8.0.4.

**Comment by pawelfus** [comment]
The line numbers does not match with http://code.highcharts.com/7.2.1/highcharts.js - could you @shakunvohra let us know the line numbers in source code? You can use http://code.highcharts.com/7.2.1/highcharts.src.js URL. Thank you.

**Comment by shakunvohra** [comment]
Line 70 in this one -  http://code.highcharts.com/7.2.1/highcharts.js 

I did find few more, I will try to compile them over the weekend.

**Comment by pawelfus** [comment]
@shakunvohra please use http://code.highcharts.com/7.2.1/highcharts.src.js URL (source), not the minified one. Thanks!

**Comment by shakunvohra** [comment]
Line 5453

**Comment by pawelfus** [comment]
Thanks!

FYI @TorsteinHonsi - this comes from here:
https://github.com/highcharts/highcharts/blob/43b54b7b8dfbe10ce7b54ee68ad115d9113fb437/ts/parts/SvgRenderer.ts#L3702-L3713

**Comment by bananabr** [comment]
Not sure if I follow what's the problem here. Can anyone come up with a PoC?

**Comment by pawelfus** [comment]
I agree, a demo with the issue would help us to understand what kind of attack we need to protect from.

@shakunvohra - could you create one? Or provide details on how to recreate it. Thanks!

**Comment by shakunvohra** [comment]
@pawelfus @bananabr I do not have much information, we use a security tool called CheckMarx and that flagged it.
As mentioned above as well - If you guys think it;s a non issue, I can make it non-exploitable in our tool, but i just wanted to bring it up in case it's an issue.


Method function at line 1302 of /js/highcharts-7.2.1/highcharts.js gets user input for
the split element. This element’s value then flows through client-side code without being properly sanitized or
validated and is eventually displayed to the user in function at line 1302 of
js/highcharts-7.2.1/highcharts.js.This may enable a DOM XSS attack.

**Comment by bananabr** [comment]
My understanding is that an attacker would only have control over the window.location search (query string) and hash. Hash is stripped at the beginning, so it's not a problem. Then renderer.url is used in 3 different places:

```js
                        }
                        // Set the reference to the gradient object
                        value = 'url(' + renderer.url + '#' + id + ')';
                        elem.setAttribute(prop, value);
```

```js
                            // label() is placed in a group, text() is standalone
                            element: this.text ? this.text.element : elem
                        });
                    }
                    // Set basic options:
                    // Use `setAttributeNS` because Safari needs this..
                    textPathElement.setAttributeNS('http://www.w3.org/1999/xlink', 'href', this.renderer.url + '#' + textPathId);
                    // Presentation attributes:
                    // dx/dy options must by set on <text> (parent),
                    // the rest should be set on <textPath>
                    if (defined(attrs.dy)) {
                        textPathElement.parentNode
```

```js
             *        The clipping rectangle. If skipped, the current clip is removed.
             *
             * @return {Highcharts.SVGElement}
             *         Returns the SVG element to allow chaining.
             */
            SVGElement.prototype.clip = function (clipRect) {
                return this.attr('clip-path', clipRect ?
                    'url(' + this.renderer.url + '#' + clipRect.id + ')' :
                    'none');
            };
            /**
             * Calculate the coordinates needed for drawing a rectangle crisply and
             * return the calculated attributes.
             *
             * @fu
```

As single quotes are stripped and setAttributeNS escapes it's input, I don't see how one could escape the attribute value and gain arbitrary JavaScript code execution.

**Comment by pawelfus** [comment]
I tried yesterday to exploit this (locally, as jsFiddle and others have their own issues) - failed, in Chrome at least. It seems it's a false positive report from CheckMarx.

Anyway, let's keep this open. If anyone will be able to provide steps to use this to attack a website, please let us know
